5-performance-optimization.ipynb

```diff
@@ -0,0 +1,764 @@
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "\n",
+    "import dask.dataframe as dd\n",
+    "\n",
+    "ddf = dd.read_csv(\n",
+    "    \"s3://coiled-datasets/uber-lyft-tlc-sample/csv-10/*\", \n",
+    "    # \"s3://coiled-datasets/uber-lyft-tlc-sample/csv-ill/*\", \n",
+    "    dtype={\"wav_match_flag\": \"category\"},  # worth adding a blurb around why we set this?\n",
+    ")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "da2e59c0",
+   "metadata": {},
+   "source": [
+    "After some initial exploration, we see that the columns representing on-scene and pickup times are stored as strings(?)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "63f2390e-5983-4844-8c37-2651afd8d940",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "\n",
+    "# Convert to datetime\n",
+    "ddf[\"on_scene_datetime\"] = dd.to_datetime(ddf[\"on_scene_datetime\"], format=\"mixed\")\n",
+    "ddf[\"pickup_datetime\"] = dd.to_datetime(ddf[\"pickup_datetime\"], format=\"mixed\")\n",
+    "\n",
+    "# Unpack columns\n",
+    "ddf = ddf.assign(\n",
+    "    accessible_vehicle=ddf.on_scene_datetime.isnull(),\n",
+    "    pickup_month=ddf.pickup_datetime.dt.month,\n",
+    "    pickup_dow=ddf.pickup_datetime.dt.dayofweek,\n",
+    "    pickup_hour=ddf.pickup_datetime.dt.hour,\n",
+    ")\n",
+    "ddf = ddf.drop(columns=[\"on_scene_datetime\", \"pickup_datetime\"])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "3b502560",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "# Format airport_fee\n",
+    "ddf[\"airport_fee\"] = ddf[\"airport_fee\"].replace(\"None\", 0).astype(float).fillna(0)\n",
+    "\n",
+    "# Remove outliers\n",
+    "lower_bound = 0\n",
+    "Q3 = ddf[\"trip_time\"].quantile(0.75)\n",
+    "upper_bound = Q3 + (1.5 * (Q3 - lower_bound))\n",
+    "ddf = ddf.loc[(ddf[\"trip_time\"] >= lower_bound) & (ddf[\"trip_time\"] <= upper_bound)]\n",
+    "\n",
+    "# # Categorize *_flag columns\n",
+    "# ddf = ddf.categorize(columns=[\"shared_request_flag\", \"shared_match_flag\", \"access_a_ride_flag\", \"wav_request_flag\", \"wav_match_flag\"])\n",
+    "\n",
+    "service_names = {\n",
+    "    \"HV0002\": \"juno\",\n",
+    "    \"HV0005\": \"lyft\",\n",
+    "    \"HV0003\": \"uber\",\n",
+    "    \"HV0004\": \"via\",\n",
+    "}\n",
+    "\n",
+    "ddf[\"service_names\"] = ddf[\"hvfhs_license_num\"].map(service_names)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "6f2b91b4-8d6a-4366-bd5d-970e4b869935",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "ddf"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "24e4e66b-24f7-4f6a-907a-96a34d48b7d0",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "Do some analysis"
```

Comment:
Would be good to expand this a bit. Maybe: "Now that the data is cleaned up, lets do some preliminary analysis. For this tutorial we are mostly interested in how long each analysis takes to run."